doc/developer/design/20231117_copy_to_s3.md

```diff
@@ -0,0 +1,304 @@
+- When they run the `COPY` command, to keep things simple we'll block the UI till the operation
+is complete, similar to any other query.
+- If successful, the output of the `COPY` command should return the number of rows written to S3
+similar to a [Postgres `COPY` output](https://www.postgresql.org/docs/current/sql-copy.html) of
+`COPY <row_count>`.
+- If the operation fails midway for some reason or is cancelled by the user, we should show
+a notice to the users that incomplete data could have been written to the S3 path and will
+need to be cleaned up.
+- Since we'll support scalar expressions in the s3 path, user can give a path like
+.`'s3://prefix/path/' || mz_now()` so that they can run the same query again and again without
+needing to update the path or clean up previous data.
+
+## Implementation
+
+### Parser
+#### S3 path
+We should accept the S3 path as an arbitrary scalar expression which we later convert to
+a string in the planner.
+```
+Uri::from_str(lower_uncorrelated().eval().unwrap_str())
+```
+
+#### `MAX FILE SIZE` Option
+We should introduce a `Memory` type which can take either an integer or a parseable string
+like `'2 GB'`, similar to the implementation for `Duration`. The multiplier should be
+1024 to keep it consistent with Postgres. If the value is not provided we should select
+an appropriate default.
+
+Snowflake uses 16MB as default for their `MAX_FILE_SIZE` whereas Redshift defaults to
+6.2GB for similar `MAXFILESIZE` parameter. Also, Snowflake recommends for data in S3 to be
+used in external tables, file sizes should be under 256MB. Given that, 256MB
+can be the default for our `MAX FILE SIZE`.
+
+### Which cluster to run things on?
+Currently existing supported `COPY ... TO STDOUT` command fetches the result to `environmentd`.
+This will not be advisable for `COPY ... TO S3` as the data size could be large. Therefore,
+we should trigger a dataflow in a `clusterd` which should eventually write to S3 using
+user's aws connection details. The next question being, should it be a Storage cluser or a
+Compute cluster (at least till the time the cluster unification is done).
+
+We should use the Compute cluster. Compute already has a concept of
+["Compute Sinks"](https://github.com/MaterializeInc/materialize/blob/v0.77.1/src/compute-types/src/sinks.rs#L100-L103),
+which include Subscribe and Persist (for materialized views). We can create another
+Compute S3 sink, which will actually read from persist and write to S3. Keeping this in a
+compute cluster also means,
+- We can easily support `SELECT` queries in a `COPY ... TO S3` command if things are
+already in compute.
+- Via the Persist Compute Sink, there's already precedence to talking to S3, we should
+be able to leverage some existing code.
+- This also will be consistent with the behaviour that a user can select the compute cluster
+where they want their query to run like they do for other queries.
```

Comment:
A thing that's missing in this discussion is that we will need to invent some new way to communicate possible sink errors, like the ones described in "User Experience" above. So far all existing sinks communicate errors by writing them into the sinked collection, but that won't be possible in this case if we want to show errors back to the user via pgwire.

Reply:
OOC what happens in case of SELECTs or SUBSCRIBEs?

Reply:
Both communicate results back to envd using `ComputeResponse`s (either `PeekResponse` or `SubscribeResponse`). And both response types have an `Error` variant that gets sent when an error is encountered along the way. Then adapter has some dedicated code to handle these error responses by converting them into pgwire errors (and, in the case of `SUBSCRIBE`s, cancelling them).

Reply:
> invent some new way to communicate possible sink errors, like the ones described in "User Experience" above

Any suggestions on what that protocol would look like, @teskje?

Reply:
Ah so similar to PeekResponse at https://github.com/MaterializeInc/materialize/blob/main/src/compute-client/src/protocol/response.rs#L210-L217, we can have something like 

```
pub enum CopyResponse {
    /// Returned number of rows copies.
    Rows(u64),
    /// Error of an unsuccessful copy.
    Error(String),
    /// The copy was canceled.
    Canceled,
}
```

Reply:
I haven't thought about this in depth, but perhaps we should introduce a new response type that resembles `PeekResponse` and has an `Ok` and an `Error` variant, to communicate the result of the copy operation. Or, if we wanted to return intermediary status updates, something like a `SubscribeResponse` would work, that continually produces updates about, e.g. the number of bytes uploaded.

Reply:
Added `CopyResponse` section https://github.com/moulimukherjee/materialize/blob/design-doc-copy-to-s3/doc/developer/design/20231117_copy_to_s3.md#copyresponse